src/templatizer/array-selector.html

```diff
@@ -0,0 +1,315 @@
+          if (idx >= start) {
+            this.unlinkPaths('selected');
+            this.unlinkPaths('selectedItem');
+            if (idx >= (start + removed)) {
+              let newItemPath = 'items.' + (idx + added - removed);
+              this.linkPaths('selected', newItemPath);
+              this.linkPaths('selectedItem', newItemPath);
+            }
+          }        
+        }
+      }
+
+      _unlinkMulti(sidx) {
+        const links = this.__dataLinkedPaths;
+        // shift other selected items down
+        for (let i=sidx+1; i<this.selected.length; i++) {
+          this.linkPaths('selected.' + (i-1), links['selected.' + i]);
+        }
+        // unlink last
+        this.unlinkPaths('selected.' + (this.selected.length-1));
+      }
+
+      /**
+       * Clears the selection state.
+       *
+       * @method clearSelection
+       */
+      clearSelection() {
+        // Unbind previous selection
+        if (Array.isArray(this.selected)) {
+          for (var i=0; i<this.selected.length; i++) {
+            this.unlinkPaths('selected.' + i);
+          }
+        } else {
+          this.unlinkPaths('selected');
+          this.unlinkPaths('selectedItem');
+        }
+        // Initialize selection
+        if (this.multi) {
+          if (!this.selected || this.selected.length) {
+            this.selected = [];
+          }
+        } else {
+          this.selected = null;
+        }
+        this.selectedItem = null;
+      }
+
+      /**
+       * Returns whether the item is currently selected.
+       *
+       * @method isSelected
+       * @param {*} item Item from `items` array to test
+       * @return {boolean} Whether the item is selected
+       */
+      isSelected(item) {
+        if (this.multi) {
+          // TODO(kschaaf) use Map for perf
+          return this.selected.indexOf(item) >= 0;
+        } else {
+          return this.selected == item;
+        }
+      }
+
+      /**
+       * Deselects the given item if it is already selected.
+       *
+       * @method deselect
+       * @param {*} item Item from `items` array to deselect
+       */
+      deselect(item) {
+        if (this.multi) {
+          let sidx = this.selected.indexOf(item);
+          if (sidx >= 0) {
+            this._unlinkMulti(sidx);
+            this.splice('selected', sidx, 1);
+          }
+        } else {
+          if (!item || this.selected === item){
+            this.selected = null;
+            this.selectedItem = null;
+            this.unlinkPaths('selected');
+            this.unlinkPaths('selectedItem');            
+          }
+        }
+      }
+
+      /**
+       * Selects the given item.  When `toggle` is true, this will automatically
+       * deselect the item if already selected.
+       *
+       * @method select
+       * @param {*} item Item from `items` array to select
+       */
+      select(item) {
+        let idx = this.items.indexOf(item);
```

Comment:
The removal of `Collection` ultimately drives this change since `collection.getKey(item)` was O(1).  One option would be to introduce a `selectIndex(idx)` and `deselectIndex(idx)` API which would avoid the item->index lookup when the index is known apriori (e.g. selection following tap on a `dom-repeat`'ed item).

When you say ' I just hit this problem in iron-list...', what do you mean?  You hit a real-world performance problem due to this particular implementation? Or a similar O(n) problem elsewhere?